agent/src/main/scala/io/scalac/mesmer/agent/Agent.scala

```diff
@@ -2,33 +2,91 @@ package io.scalac.mesmer.agent
 
 import java.lang.instrument.Instrumentation
 
+import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation
+import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer
 import net.bytebuddy.agent.builder.AgentBuilder
+import net.bytebuddy.description.`type`.TypeDescription
+import net.bytebuddy.dynamic.DynamicType
+import net.bytebuddy.matcher.ElementMatcher
 import org.slf4j.LoggerFactory
 
-import io.scalac.mesmer.agent.Agent.LoadingResult
+import scala.jdk.CollectionConverters.SeqHasAsJava
 
+import io.scalac.mesmer.agent.util.i13n
+import io.scalac.mesmer.agent.util.i13n.InstrumentationDetails
+
+final case class Agent private (private[agent] val instrumentations: Set[AgentInstrumentation]) extends {
+  import Agent._
+
+  def ++(other: Agent): Agent = Agent(instrumentations ++ other.instrumentations)
+
+  def plus(other: Agent): Agent = ++(other)
+
+  def ++(other: AgentInstrumentation): Agent = Agent(instrumentations + other)
+
+  def installOnMesmerAgent(builder: AgentBuilder, instrumentation: Instrumentation): LoadingResult = {
+
+    def mesmerAgentInstallation(agentInstrumentation: AgentInstrumentation): LoadingResult = {
+      val typeInstrumentation: i13n.TypeInstrumentation     = agentInstrumentation.typeInstrumentation
+      val instrumentationDetails: InstrumentationDetails[_] = typeInstrumentation.`type`.name
+      builder
+        .`type`(typeInstrumentation.`type`.desc)
+        .transform((underlying, _, _, _) => typeInstrumentation.transformBuilder(underlying))
+        .installOn(instrumentation)
+      if (instrumentationDetails.isFQCN) LoadingResult(instrumentationDetails.name +: agentInstrumentation.load)
+      else LoadingResult.empty
+    }
+
+    // Sorting a set is very brittle when it comes to determining the installation order.
+    // See more: https://github.com/ScalaConsultants/mesmer-akka-agent/issues/294
+    instrumentations.toSeq.sorted
+      .map(mesmerAgentInstallation)
+      .fold(LoadingResult.empty)(_ ++ _)
+  }
+
+  def asOtelTypeInstrumentations: java.util.List[TypeInstrumentation] = {
```

Comment:
That's surely a subject to change - I took a shortcut there. Ideally:
 - the otel related code should be in otel-extension package
 - the instrumentations module should keep the common code for instrumenting both mesmer agent and otel agent
 - the mesmer agent code should keep only mesmer agent specific code in it.
 

 Note that: 
  - we need to put the `instrumentation` module on the instrumented application's classpath in order for bytebuddy be able to actually instrument the classes (otherwise we get `classNotFoundException`s)
  - the `otel-extension` should not depend on agent - that's my mistake but it can be solved after I refactor that. I didn't want to make a huge pr but it surely deserves an issue now:

Reply:
Ok. Now I see what you meant here: 

> If otel-extension has to depend on agent anyway we can leave all instrumentations there and maintain the same level of flexibility - WDYT?

indeed we do not need instrumentations. Initially I thought that some code is going to be mesmer agent-specific so in my head it made sense to divide it as described above. It turns out that there's very few code like this so now I agree with you. 

I will remove the "instrumentations" module for now and leave agent and otel-extension only. Once we remove the mesmer agent (if ever) we can rename the module to "instrumentations" because that's what essentially it will contain (I think).

Reply:
fixed back to otel-extension + agent modules only

Reply:
If you're ok with this, let's do the cleanup in a separate pr for: https://github.com/ScalaConsultants/mesmer-akka-agent/issues/306. It would require extracting TypeInstrumentation from package object and some other restructuring so i'd prefer to do that as a separate chunk of work later